File: code-comments/simpson.c
// BSD License
#include "LICENSE.h"

#include <math.h>
#include <stdio.h>

double integrate(double (*f)(double), double a, double b, double n) {
    double h = (b - a) / n; // Interval width
    double sum = 0.0;       // Accumulator
    sum += f(a);            // First term
    double tmp = 0.0;       // Even terms
    for (int j = 1; j <= floor(n / 2 - 1); j += 1) {
        tmp += f(a + 2 * j * h);
    }
    sum += 2.0 * tmp;
    tmp = 0.0;              // Odd terms
    for (int j = 1; j <= floor(n / 2); j += 1) {
        tmp += f(a + (2 * j - 1) * h);
    }
    sum += 4.0 * tmp;
    sum += f(b);            // Last term
    sum *= h / 3.0;
    return sum;
}

double g(double x) { return exp(-x * x); }

int main(void) {
    double p = integrate(g, -10.0, 10.0, 100);
    double 𝛑 = p * p;
    printf("𝛑 = %22.20lf (𝚫 = %22.20lf)\n", 𝛑, M_PI - 𝛑);
    return 0;
}

File: W-22/Examples/python/simpson.py
#!/usr/bin/env python3

def even(n): return n % 2 == 0

def integrate(f, a, b, n):
    h = (b - a) / n
    sum = f(a) + f(b)
    for j in range(1, n):
        sum += 2.0 * f(a + j * h) if even(j) else 4.0 * f(a + j * h)
    sum *= h / 3.0
    return sum

from math import exp as exp

def g(x): return exp(-x * x)

from math import sqrt as sqrt

def h(x): return sqrt(1 - x * x)

def l(x): return sqrt(x) * exp(-x)

from math import pi as pi

if __name__ == '__main__':
    p = integrate(g, -7.0, 7.0, 100)
    𝛑 = p * p
    print(f"𝛑 = {𝛑} (𝚫 = {pi - 𝛑})")
    𝛑 = 2 * integrate(h, -1.0, 1.0, 5000000)
    print(f"𝛑 = {𝛑} (𝚫 = {pi - 𝛑})")
    p = integrate(l, 0.0, 100.0, 10000000)
    𝛑 = 4 * p * p
    print(f"𝛑 = {𝛑} (𝚫 = {pi - 𝛑})")
